diff --git a/third_party/zlib/contrib/arm/inflate.c b/third_party/zlib/contrib/arm/inflate.c
index e40322c..74c5541 100644
--- a/third_party/zlib/contrib/arm/inflate.c
+++ b/third_party/zlib/contrib/arm/inflate.c
@@ -229,6 +229,8 @@ int stream_size;
     state->strm = strm;
     state->window = Z_NULL;
     state->mode = HEAD;     /* to pass state test in inflateReset2() */
+    /* 1L is the result of adler32() on zero length data */
+    state->check = 1L;
     ret = inflateReset2(strm, windowBits);
     if (ret != Z_OK) {
         ZFREE(strm, state);
diff --git a/third_party/zlib/inflate.c b/third_party/zlib/inflate.c
index 69b769a..7631ba1 100644
--- a/third_party/zlib/inflate.c
+++ b/third_party/zlib/inflate.c
@@ -228,7 +228,8 @@ int stream_size;
     state->strm = strm;
     state->window = Z_NULL;
     state->mode = HEAD;     /* to pass state test in inflateReset2() */
-    state->check = adler32(0L, Z_NULL, 0);
+    /* 1L is the result of adler32() on zero length data */
+    state->check = 1L;
     ret = inflateReset2(strm, windowBits);
     if (ret != Z_OK) {
         ZFREE(strm, state);
