diff --git a/source/i18n/digitlst.cpp b/source/i18n/digitlst.cpp
index a3b3406..ab5b557 100644
--- a/source/i18n/digitlst.cpp
+++ b/source/i18n/digitlst.cpp
@@ -60,8 +60,8 @@
 # endif
 #endif
 
-#if U_USE_STRTOD_L && !U_PLATFORM_USES_ONLY_WIN32_API
-# if U_PLATFORM == U_PF_CYGWIN
+#if U_USE_STRTOD_L
+# if U_PLATFORM_USES_ONLY_WIN32_API || U_PLATFORM == U_PF_CYGWIN
 #   include <locale.h>
 # else
 #   include <xlocale.h>
