Index: source/common/ucasemap.cpp
===================================================================
--- source/common/ucasemap.cpp	(revision 39499)
+++ source/common/ucasemap.cpp	(revision 39500)
@@ -200,7 +200,7 @@
         return -1;  // integer overflow
     }
     int32_t limit=destIndex+length;
-    if(limit<destCapacity) {
+    if(limit<=destCapacity) {
         U8_APPEND_UNSAFE(dest, destIndex, c);
     }
     return limit;
