diff --git a/source/common/unicode/uclean.h b/source/common/unicode/uclean.h
index 5beb1f7c..f83dd5ef 100644
--- a/source/common/unicode/uclean.h
+++ b/source/common/unicode/uclean.h
@@ -149,7 +149,7 @@ typedef void  U_CALLCONV UMemFreeFn (const void *context, void *mem);
  *  @system
  */  
 U_STABLE void U_EXPORT2 
-u_setMemoryFunctions(const void *context, UMemAllocFn * U_CALLCONV a, UMemReallocFn * U_CALLCONV r, UMemFreeFn * U_CALLCONV f, 
+u_setMemoryFunctions(const void *context, UMemAllocFn * a, UMemReallocFn * r, UMemFreeFn * f, 
                     UErrorCode *status);
 
 U_CDECL_END
