diff -ru protobuf2/src/google/protobuf/descriptor.h protobuf/src/google/protobuf/descriptor.h
--- protobuf2/src/google/protobuf/descriptor.h	2017-07-06 15:32:27.892104861 -0700
+++ protobuf/src/google/protobuf/descriptor.h	2017-07-06 15:19:07.042709476 -0700
@@ -54,6 +54,7 @@
 #ifndef GOOGLE_PROTOBUF_DESCRIPTOR_H__
 #define GOOGLE_PROTOBUF_DESCRIPTOR_H__
 
+#include <map>
 #include <memory>
 #ifndef _SHARED_PTR_H
 #include <google/protobuf/stubs/shared_ptr.h>
